### ISSUE TITLE
Different servers listening on `0.0.0.0` and `127.0.0.1` results in ignored `server_name` and `default_server`.

### ISSUE THREAD

**Issue by Timshel** [nginx/nginx#220]
### Environment

Include the result of the following commands:
  - `nginx version: nginx/1.27.1`
  - `6.9.7+bpo-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.9.7-1~bpo12+1 (2024-07-03) x86_64 GNU/Linux`

### Description

With multiple server block listening on `0.0.0.0`  adding one listening on a specific ip ex `127.0.0.1` will result in other servers being ignored for queries on `127.0.0.1`.

- [x] The bug is reproducible with the latest version of nginx
- [x] The nginx configuration is minimized to the smallest possible
to reproduce the issue and doesn't contain third-party modules

#### nginx configuration

```
server {
    listen 0.0.0.0:80 default_server;
    server_name _;
    location / {
        return 403;
    }
}

server {
    listen 0.0.0.0:80;
    server_name hello.lan;
    location / {
        default_type application/text;
        return 200 'Hello';
    }
}

server {
    listen 127.0.0.1:80;
    server_name bug.lan;
    location / {
        default_type application/text;
        return 200 'Bug ?';
    }
}

```
When run with
```bash
> docker run -it -d --rm --network host -v conf:/etc/nginx/conf.d/test.conf  nginx:1.27.1
> curl http://hello.lan --resolve hello.lan:80:127.0.0.1
 Bug ?
> curl http://toto.lan --resolve toto.lan:80:127.0.0.1
 Bug ?
```

As opposed to when run without the last `server` listening on `127.0.0.1:80`:
```bash
> curl http://hello.lan --resolve hello.lan:80:127.0.0.1
 Hello
> curl http://toto.lan --resolve toto.lan:80:127.0.0.1
<html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
<hr><center>nginx/1.27.1</center>
</body>
</html>
```

**Comment by jo-carter** [comment]
You can't bind to INADDR_ANY(0.0.0.0) and specific address on the same port, with Linux. Check your nginx error log you should see there is in fact an emerg error repeated related to failed binding in a similar fashion to [this](https://trac.nginx.org/nginx/ticket/1457)

**Comment by Timshel** [comment]
I see no such error, the whole docker log:

```bash
docker run -it --rm --network host -v conf:/etc/nginx/conf.d/test.conf  nginx:1.27.1
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
10-listen-on-ipv6-by-default.sh: info: Getting the checksum of /etc/nginx/conf.d/default.conf
10-listen-on-ipv6-by-default.sh: info: Enabled listen on IPv6 in /etc/nginx/conf.d/default.conf
/docker-entrypoint.sh: Sourcing /docker-entrypoint.d/15-local-resolvers.envsh
/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
/docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
/docker-entrypoint.sh: Configuration complete; ready for start up
2024/09/30 08:02:42 [notice] 1#1: using the "epoll" event method
2024/09/30 08:02:42 [notice] 1#1: nginx/1.27.1
2024/09/30 08:02:42 [notice] 1#1: built by gcc 12.2.0 (Debian 12.2.0-14) 
2024/09/30 08:02:42 [notice] 1#1: OS: Linux 6.9.7+bpo-amd64
2024/09/30 08:02:42 [notice] 1#1: getrlimit(RLIMIT_NOFILE): 1048576:1048576
2024/09/30 08:02:42 [notice] 1#1: start worker processes
2024/09/30 08:02:42 [notice] 1#1: start worker process 29
2024/09/30 08:02:42 [notice] 1#1: start worker process 30
2024/09/30 08:02:42 [notice] 1#1: start worker process 31
2024/09/30 08:02:42 [notice] 1#1: start worker process 32
2024/09/30 08:02:42 [notice] 1#1: start worker process 33
2024/09/30 08:02:42 [notice] 1#1: start worker process 34
2024/09/30 08:02:42 [notice] 1#1: start worker process 35
2024/09/30 08:02:42 [notice] 1#1: start worker process 36
2024/09/30 08:02:42 [notice] 1#1: start worker process 37
2024/09/30 08:02:42 [notice] 1#1: start worker process 38
2024/09/30 08:02:42 [notice] 1#1: start worker process 39
2024/09/30 08:02:42 [notice] 1#1: start worker process 40
2024/09/30 08:02:42 [notice] 1#1: start worker process 41
2024/09/30 08:02:42 [notice] 1#1: start worker process 42
2024/09/30 08:02:42 [notice] 1#1: start worker process 43
2024/09/30 08:02:42 [notice] 1#1: start worker process 44
127.0.0.1 - - [30/Sep/2024:08:03:12 +0000] "GET / HTTP/1.1" 200 5 "-" "curl/7.88.1" "-"
127.0.0.1 - - [30/Sep/2024:08:03:27 +0000] "GET / HTTP/1.1" 200 5 "-" "curl/7.88.1" "-"
```

I checked that if I have another nginx running then bind errors are reported in the docker log (so not an issue of incomplete log).

Netstat report that nginx is bound on `0.0.0.0:80` as expected.
And the three virtual server config appear "active", if I curl on the external ip then I'll either have the `403` or `Hello`.

**Comment by jo-carter** [comment]
Ah yes, my mistake - that issue only occurs by default during a runtime (reload) switch from listening on 0.0.0.0/0 to a non-wildcard address on the same port.

Nginx skips listen directives for all other addresses on a given port if that port is used with wildcard listen address within the nginx configuration - see the [bind parameter of the listen directive](https://nginx.org/en/docs/http/ngx_http_core_module.html#listen) which disables this behavior.

>bind
>instructs to make a separate bind() call for a given address:port pair. This is useful because if there are several listen directives >with the same port but different addresses, and one of the listen directives listens on all addresses for the given port (*:port), >nginx will bind() only to *:port. It should be noted that the getsockname() system call will be made in this case to determine the >address that accepted the connection. If the setfib, fastopen, backlog, rcvbuf, sndbuf, accept_filter, deferred, ipv6only, reuseport, >or so_keepalive parameters are used then for a given address:port pair a separate bind() call will always be made.

Using the bind parameter will not resolve your issue, as per the issue previously linked to the Linux Kernel does not support the desired behavior. You will see bind errors now however.

**Comment by Timshel** [comment]
> Nginx skips listen directives for all other addresses on a given port if that port is used with wildcard listen address within the nginx configuration

Yes I would expect it.

But from the bind documentation you linked:
> It should be noted that the getsockname() system call will be made in this case to determine the >address that accepted the connection.

So I would expect it would then correctly apply each virtual server config using this address.

**Comment by arut** [comment]
In nginx servers are grouped by listen address. Whenever you connect to a particular address, server is chosen within this group only. I don't think we want to change this.

**Comment by Timshel** [comment]
I'm not sure, I understand your point, when running nginx the only listen address is `0.0.0.0:80`.
The `127.0.0.1:80` listen is ignored since as mentioned you can't bind twice, and it's in essence a subset of the wildcard bind.

When looking at the code there is two virtual listen address groups `0.0.0.0:80` and `127.0.0.1:80` but since those are virtual I don't understand why the group `127.0.0.1:80` should not include the two server definition from the wildcard servers (It's in essence what I changed in the PR).

**Comment by arut** [comment]
In nginx these are separate groups and we have no plans to change this. If we change this, it would mess up existing configurations. Also, it could break them in case there's the same server name in both groups. I believe it's possible to work around this in your configuration.

**Comment by Timshel** [comment]
> Also, it could brake them in case there's the same server name in both groups

Yes this would be an invalid config (if it does not break at the moment not sure how it's better).

> I believe it's possible to work around this in your configuration.

It is, but it's a pain since it means I'm forced to replace all wildcard bind ...

> In nginx these are separate groups and we have no plans to change this.

Ok but in this case I believe this config should output an error, because I was quite surprised when I added the specific ip bind to a configuration with lots of wildcard binds and suddenly all routing for traffic relying on the latter stopped ...

**Comment by Timshel** [comment]
...

**Comment by janhaa** [comment]
Hi, we just stumbled upon this issue and https://github.com/nginx/nginx/issues/220#issuecomment-2474041921 resolved it for us, but seems a bit counterintuitive on first encounter, possibly this could be explictly explained in the docs here: https://nginx.org/en/docs/http/request_processing.html
